draft-pda-core.md

```diff
@@ -291,11 +293,427 @@ Each PA task in this document is divided into three sub-protocols as follows.
+   parameters used for each task.
+
+[TODO: It would be clearer to include a "pre-conditions" section prior to each
+"phase" of the protocol.]
+
+## Upload {#pa-upload}
+
+[TODO: Add an illustration of this sub-protocol.]
+
+Uploading a report involves two requests to the leader. In the  *upload start
+request*, the client discovers the protocol-specific parameters it needs to
+generate the report, as well as the endpoint URL of each helper. In the *upload
+finish request*, it uploads its report to the leader.
+
+[NOTE: @acmiyaguchi pointed out that the use of an anonymizing proxy for
+uploading shares might be easier to implement if the "upload" phase involved a
+single HTTP request. However, OHTTP
+(https//www.ietf.org/archive/id/draft-thomson-http-oblivious-01.html) allows
+clients to make multiple requests through a proxy, so these kinds of use cases
+should work.]
+
+[NOTE: @cjpatton: Breaking the upload phase into two requests is useful for
+supporting Prio-like proof systems in which the leader sends the client a
+"challenge" it uses to generate the proof. (See the SIMD construction of
+{{BBCp19}}, Section 5.2. Here, the "challenge" is a randomly generated field
+element.) Because the client/leader are using TLS, another way to implement this
+would be to export a secret from the key schedule.]
+
+### Upload Start
+
+The client sends a POST request to `[leader]/upload_start` with the following
+message:
+
+~~~
+struct {
+  PATask task;
+} PAUploadStartReq;
+~~~
+
+The `task` field corresponds to the PA task for which a report will be
+generated.
+
+The leader responds to well-formed requests to `[leader]/upload_start` with
+status 200 and the following message:
+
+~~~
+struct {
+  PAParam param;
+  Url helper_urls<1..2^16-1>;
+  select (PAUploadStartResp.param.proto) {
+    case prio: PrioUploadStartResp;
+    case hits: HitsUploadStartResp;
+  }
+} PAUploadStartResp;
+
+opaque Url<1..2^16-1>;
+~~~
+
+The message includes the URL of each helper and any protocol-specific parameters
+the client needs to generate its report. The leader's response to malformed
+requests is specified in {{pa-error-common-aborts}}.
+
+### Upload Finish
+
+For each URL `[helper]` in `PAUploadStartResp.helper_urls`, the client sends a
+GET request to `[helper]/key_config`. The helper responds with status 200 and an
+`HpkeConfig` message. Next, the client collects the set of helpers it will
+upload shares to. It ignores a helper if:
+* the client and helper failed to establish a secure, helper-authenticated
+  channel;
+* the GET request to the helper URL failed or didn't return a valid key config;
+  or
+* the key config specifies a KEM, KDF, or AEAD algorithm the client doesn't
+  recognize.
+
+If the set of supported helpers is empty, then the client aborts and alerts the
+leader with "no supported helpers". Otherwise, for each supported helper the
+client issues a POST request to `[leader]/upload_finish` with a payload
```

Comment:
> Does this mean we only support 2-way secret sharing? I think that's OK, but we should make that constraint of the system more obvious.

Indeed -- this is limited to the two-server case. 

> Since all the reports are going to the same leader server, could they be combined into a single message, and hence a single request with a single state to be managed by the client and leader?

I don't see why not, though this seems like an optimization more than a functional change. 

> Is the client expected to secret share the input differently for each PAUploadFinishReq it constructs? If not, then PAUploadFinishReq.leader_share will be identical in each message, and since the leader share will be much larger than the helper share (at least in Prio where the helper's share is a PRNG seed), we should work out a way to absolve the client (which might be on an expensive and slow cellular network) of having to upload the leader share k times. 

Yes, I think that's expected. (That is, I assume the proof is fresh for each upload.)

Reply:
I think this is fine, so long as each of the `k` protocol runs the client initiates for a given report are totally independent from one another.